text/0240-app-telemetry-schema-vision-roadmap.md

```diff
@@ -0,0 +1,305 @@
+multiple times in different signals. It is expected to be a very common pattern
+to reuse the same subset of attributes or metrics across several signals.
+
+A dedicated OTEP will precisely define the structure and format of the Resolved
+Telemetry Schema.
+
+## Roadmap
+
+To facilitate the review process and progressively deliver value to the
+OpenTelemetry community, a series of OTEPs and tools are proposed.
+
+### OTEPs
+
+* Telemetry Schema versions (incremental steps):
+  * V1.1.1: Add a section to import one or several Semantic Convention
+  Registries. Once implemented, the official Semantic Convention Registry will
+  be referenced from the official OpenTelemetry Root Telemetry Schema.
+  * V1.2: Add references to parent schemas and define a schema section to
+  describe application/library signals. In this version, only metrics, logs, and
+  spans signals are described. This is the first version of the telemetry schema
+  that can be used by the Client SDK generator. In this version only one schema
+  can be inherited. More complex hierarchies will be supported in version 1.5.
+  * V1.3: Add support for events in the schema section. This OTEP is contingent
+  upon the approval of events in OpenTelemetry.
+  * V1.4: Add support for metric_groups (multivariate metrics) in the schema
+  section. This will only be relevant if a support for multivariate metrics is
+  planned in the Client SDK Generator.
+  * V1.5: Add support for multi-inheritance and for schema hierarchies of more
+  than 2 levels. This is useful for supporting complex environments and will
+  allow the distribution of semantic convention registries across vendors and
+  enterprises. This OTEP depends on the “Schema Child/Parent relationship
+  resolution” OTEP.
+* Resolved Schema Format v1.0.
+* Schema-driven Client SDK Generator.
+* Schema Child/Parent relationship resolution.
+
+### Tools
+
+OpenTelemetry tools
+
+* Semantic Convention Registry Merger (Johannes Tax's proposal): This tool will
+merge all the semantic convention files of a registry into a single and
+referencable file.
+* Search Engine: This tool will provide search capability within semantic
+convention registries and telemetry schemas.
+  * V1 focusing on the semantic convention registry.
+  * V2 focusing on telemetry schema v1.2.
+* Telemetry Schema Resolver: This tool will resolve the telemetry schema and its
+dependencies. The main outcome of this tool is a Resolved Telemetry Schema.
+* Schema-driven Client SDK Generator: This tool will generate a client SDK or a
+client API from a telemetry schema v1.2. It will be designed to be extensible,
+allowing each language SIG and the general community to create templates for
+generating client SDKs and APIs in specific programming languages.
+* Telemetry Schema Compatibility Checker: This tool will verify that two
+successive versions of the same Telemetry Schema adhere to the specified
+compatibility rules.
```

Comment:
Would be good to have some sort of component that could validate that incoming telemetry conforms to the telemetry schema it uses to describe itself. For a while, I've imagined a collector processor which inspects data and says whether or not the data conforms to the semantic conventions of its schemaUrl. This effort could be a good chance to build such a tool.

Generally, I do think its important that we have at least some sort of tool that consumes telemetry schemas. Building this tool confirms that the concepts in telemetry schema are suitable to build on top of, and acts as a reference implementation for how other consumers of telemetry schemas might work.

Reply:
Indeed, one of the objectives of this project is to be able to validate incoming telemetry among other things (e.g., privacy policy enforcement, dashboard generation, database schema generation, schema-driven data transformation, ...).

I have just made public this [repo](https://github.com/f5/otel-weaver) containing a tool (PoC) that explores and implements some aspects of this OTEP.